File: Exercise_4/C++/B.cpp
#include "LinkedList.hpp"
#include "B.hpp"
#include <ctime>
#include <iostream>

B::B() {
    node_list = new LinkedList();
}

B::~B() {
    while(node_list->getSize())
        node_list->deleteFirst();
    delete node_list;
}


void B::benchmark(int operations) {
    int i=0;
    for (i = 0; i < operations; i++) {
        Node *new_node = new Node();
        new_node->next = NULL;
        node_list->add(new_node);
        node_list->get(42);
    }
}

void B::init(int size) {
    int i=0;
    for (i=0; i < size; i++) {
        Node *new_node = new Node();
        new_node->next = NULL;
        node_list->add(new_node);
    }
}

int main() {
    B *b = new B();
    int operations = 50500000;
    std::cout << "Initializing..." << std::endl;  
    b->init(70000000);
    std::cout << "Benchmarking " << operations << " Operations. " << std::endl;  
    clock_t begin = clock();
    b->benchmark(operations);
    clock_t end = clock();  
    delete(b);
    double elapsed_secs = double(end - begin) / CLOCKS_PER_SEC;
    std::cout << "Time Elapsed  (secs) : " << elapsed_secs << std::endl;  
    return 0;
}

File: Exercise_4/Java/B.java
import java.util.Random;

/**
 * Created by milonaki on 22/12/2015.
 */
public class B {

    private final LinkedList nodeList;

    public B() {
        nodeList = new LinkedList();
    }

    public void init(int size) {
        for (int i=0; i< size; i++) {
            Node node = new Node();
            node.next = null;
            nodeList.add(node);
        }
    }


    private void benchmark(int operations) {
        Random random = new Random();
        for (int i = 0; i < operations; i++) {
            Node newNode = new Node();
            newNode.next = null;
            nodeList.add(newNode);
            nodeList.get(42);
        }
    }

    static long getFreeMemory () {
        // waits for free memory measurement to stabilize
        long init = Runtime.getRuntime().freeMemory(), init2;
        int count = 0;
        do {
            System.out.println("waiting..." + init);
            System.gc();
            try { Thread.sleep(250); } catch (Exception x) { }
            init2 = init;
            init = Runtime.getRuntime().freeMemory();
            if (init == init2) ++ count; else count = 0;
        } while (count < 5);
        System.out.println("ok..." + init);
        return init;
    }

    public static void main(String[] args) {
        B b = new B();
        System.out.println("Initializing...");
        b.init(70000000);
        long startTime = System.currentTimeMillis();
        int operations = 50500000;
        System.out.println("Benchmarking " + operations + " Operations.");
        b.benchmark(operations);
        long endTime = System.currentTimeMillis();
        System.out.println("Time elapsed (secs): " + (endTime - startTime)/1000);
    }
}
